Contract verification

barabanovro opened the pull request:
Closes #892 

## Introduced changes

This PR builds upon the changes introduced in PR #875. The original PR is still open because the Voyager APIs are not ready yet, but we require the verification feature for the Walnut Debugger. To accommodate this, we have incorporated the changes from PR #875 and made the following additions:
* Introduced the `WalnutVerificationInterface`.
* Resolved branch conflicts.
* Made minor modifications to ensure compatibility.

Furthermore, we have temporarily removed the Voyager and Starkscan interfaces since their respective APIs are not currently available.

This Pull Request adds the verify command for the contract verification. More about the details of the contract verification can found [here](https://github.com/foundry-rs/starknet-foundry/blob/master/design_documents/contract_verification.md).

* Added `verify` command to the `sncast` tool
* Modified `cast/main.rs` & created `cast/src/starknet_commands/verify.rs` mainly

## Checklist

- [x] Linked relevant issue
- [ ] Updated relevant documentation
- [ ] Added relevant tests
- [x] Performed self-review of the code
- [ ] Added changes to `CHANGELOG.md`

barabanovro commented:
Hey @Arcticae 👋
I am asking for a preliminary review, mainly involving the automated tests — could you shed more light on how to approach them?

Arcticae commented on crates/sncast/src/response/structs.rs:
```diff
@@ -144,3 +144,17 @@ pub struct TransactionStatusResponse {
 
 impl CommandResponse for TransactionStatusResponse {}
+
+#[derive(Serialize)]
```
It's just a `Result`, isn't it?

barabanovro replied:
I have removed the `verification_status` field, as you pointed out that it's similar to `Result<>`. Now, `VerifyResponse` allows us to easily implement the `CommandResponse` trait, maintaining consistency with other response structures such as `ScriptInitResponse`.

Arcticae commented on crates/sncast/src/response/structs.rs:
```diff
@@ -144,3 +144,17 @@ pub struct TransactionStatusResponse {
+    Error,
+}
+
+#[derive(Serialize)]
+pub struct VerifyResponse {
+    pub verification_status: VerificationStatus,
+    pub message: Option<String>,
+}
```
Should be probably an alias of `Result<String>` - i think that's the intent here?

Arcticae commented on crates/sncast/src/starknet_commands/verify.rs:
```diff
@@ -89,70 +78,30 @@ impl VerificationInterface for VoyagerVerificationInterface {
-        };
-        explorer_url.to_string()
+            "mainnet" => Ok("http://api.walnut.dev/v1/sn_main/verify"),
+            "goerli" => Ok("https://api.walnut.dev/v1/sn_goerli/verify"),
```
What about sepolia?

Arcticae commented on crates/sncast/src/starknet_commands/verify.rs:
```diff
@@ -89,70 +78,30 @@ impl VerificationInterface for VoyagerVerificationInterface {
-            _ => "https://testnet.starkscan.co/",
-        };
-        explorer_url.to_string()
+            "mainnet" => Ok("http://api.walnut.dev/v1/sn_main/verify"),
```
Why it doesn't have `https`?

Arcticae reviewed (commented):
The code looks mostly OK to me, 
As for the tests, you either need to:
- Construct some mock APIs and test with that
- Expose some testing APIs on your side (i would rather not depend on external APIs though since we can't control those)

Also documentation needs to be added as for any other command, in appendix. 
You can also add a guide for visibility.

barabanovro commented on crates/sncast/src/main.rs:
```diff
@@ -171,17 +171,35 @@ fn main() -> Result<()> {
+    };
+
+    let provider = get_provider(&config.url)?;
+    verify_and_warn_if_incompatible_rpc_version(&provider, &config.url).await?;
```
I have reverted this change.

However, the motivation for the initial change was that the 'verify' command does not require the RPC URL to be set. The same applies to the next comment regarding the match statement.

Arcticae replied:
Understood, we are going to address this in another issue, we're aware that this is sub-optimal

barabanovro commented:
I've addressed the PR comments and implemented tests using mock APIs with wiremock. I have used the WALNUT_API_URL env variable to pass the mock server URL to the verify command.

I'll be adding the documentation today.

barabanovro commented:
Hey @Arcticae, I’ve fixed the lint errors

Arcticae reviewed (commented):
It looks fine code-wise but i have a few security-related questions:

1. Do we need to actually submit all the code to the server? (If we do we should issue a warning for the user and make him confirm the choice, something like “You are about to sumbit your code to the 3rd party chosen verifier at <URL>, are you sure? (y/n)"
2. What exactly happens to the code after it's submitted to the server? I think that it's not clear right now, what will be displayed about it on the explorer exactly? Will the code be displayed to all the users?

barabanovro commented:
1. Yes, we need to submit all of the code. This way, it doesn't depend on the Scarb version installed locally.
2. The warning for the user can be: "You are about to submit your code to the third-party chosen verifier at Walnut, and the code will be publicly available through Walnut's APIs. Are you sure? (y/n)" so it’s clear that the source code can be displayed to all users.

If this is okay with you, then I can implement the warning.

barabanovro commented on crates/sncast/src/starknet_commands/verify.rs:
```diff
@@ -0,0 +1,163 @@
+
+        // Create the JSON payload with "contract name," "address," and "source_code" fields
+        let payload = VerificationPayload {
+            contract_name: contract_name.to_string(),
```
Yes, it is currently validated on the server. I will add client-side validation since it's easy to do

Arcticae commented:
> 1. Yes, we need to submit all of the code. This way, it doesn't depend on the Scarb version installed locally.
> 2. The warning for the user can be: "You are about to submit your code to the third-party chosen verifier at Walnut, and the code will be publicly available through Walnut's APIs. Are you sure? (y/n)" so it’s clear that the source code can be displayed to all users.
> 
> If this is okay with you, then I can implement the warning.

Yes please do that - and also say that it's whole workspace's code, that part is important as well i think

barabanovro commented:
All PR comments have been addressed

rjnrohit commented:
RPC url is not need for contract verification process
but 
[`let provider = get_provider(&config.url)?;`](https://github.com/foundry-rs/starknet-foundry/blob/b66f1879d54c801b9d28d05a966def27f91c7815/crates/sncast/src/main.rs#L163) will throw an error if we don't pass the rpc url.

Arcticae commented on crates/sncast/src/starknet_commands/verify.rs:
```diff
@@ -0,0 +1,184 @@
+
+    /// Assume "yes" as answer to confirmation prompt and run non-interactively
+    #[clap(long, default_value = "false")]
+    pub yes: bool,
```
This option should be named differently, i think it's important to let the user know what he's confirming. There's only one prompt now, so we can name this precisely how we want, since we know which confirmation we want to skip here

Arcticae commented on crates/sncast/tests/e2e/verify.rs:
We lack a test with no consent after the prompt

barabanovro replied:
We have the test case [here](https://github.com/walnuthq/starknet-foundry/blob/37f40f1a9d39c18e9965888d1042d9aea15605e2/crates/sncast/tests/e2e/verify.rs#L110). Additionally, I have added a test case that includes the confirm verification flag. Please let me know if I misunderstood the comment.

ksew1 commented on crates/sncast/src/starknet_commands/verify.rs:
```diff
@@ -0,0 +1,184 @@
+    pub contract_name: String,
+
+    /// Block explorer to use for the verification
+    #[clap(short = 'v', long = "verifier", value_parser = ["walnut"], default_value = "walnut")]
```
```suggestion
    #[clap(short, long, value_parser = ["walnut"], default_value = "walnut")]
```

ksew1 commented on crates/sncast/src/starknet_commands/verify.rs:
```diff
@@ -0,0 +1,184 @@
+    pub verifier: String,
+
+    /// The network on which block explorer will do the verification
+    #[clap(short = 'n', long = "network", value_parser = ["mainnet", "sepolia"])]
```
```suggestion
    #[clap(short, long, value_parser = ["mainnet", "sepolia"])]
```

ksew1 commented on crates/sncast/src/starknet_commands/verify.rs:
```diff
@@ -0,0 +1,184 @@
+
+    /// Block explorer to use for the verification
+    #[clap(short = 'v', long = "verifier", value_parser = ["walnut"], default_value = "walnut")]
+    pub verifier: String,
```
Let's use an enum here, then we won't need to have `_ =>` branches, which are unreachable. We use this in account add for `account_type`

ksew1 commented on crates/sncast/src/starknet_commands/verify.rs:
```diff
@@ -0,0 +1,184 @@
+
+    /// The network on which block explorer will do the verification
+    #[clap(short = 'n', long = "network", value_parser = ["mainnet", "sepolia"])]
+    pub network: String,
```
Same suggestion as for `verifier`

ksew1 commented on docs/src/appendix/sncast/verify.md:
```diff
@@ -0,0 +1,25 @@
+The name of the contract. The contract name is the part after the `mod` keyword in your contract file.
+
+## `--verifier, -v <VERIFIER>`
+Required.
```
```suggestion
Optional.
```
This is optional, as there is a default value.

ksew1 commented on docs/src/appendix/sncast/verify.md:
Let's add also entry for `--yes`

ksew1 commented on crates/sncast/src/starknet_commands/verify.rs:
```diff
@@ -0,0 +1,184 @@
+    pub network: String,
+
+    /// Assume "yes" as answer to confirmation prompt and run non-interactively
+    #[clap(long, default_value = "false")]
```
```suggestion
    #[clap(short, long, default_value = "false")]
```

barabanovro replied:
I haven't implemented this change because I renamed `--yes` to `--confirm-verification` and now `--contract-name`, `--confirm-verification` both flags could share the same short argument name. Additionally, Tomasz emphasized the importance of clearly informing the user about what they are confirming, so I opted to leave it as is for now. Please advise if any changes are required.

ksew1 replied:
It now makes sense not to add a short option.

ksew1 reviewed (commented):
Great job! 😄 Couple small suggestions.

rjnrohit commented:
A `package` variable must be added inside the `Verify` struct, refer `Declare` object

ksew1 commented:
> RPC url is not need for contract verification process but [`let provider = get_provider(&config.url)?;`](https://github.com/foundry-rs/starknet-foundry/blob/b66f1879d54c801b9d28d05a966def27f91c7815/crates/sncast/src/main.rs#L163) will throw an error if we don't pass the rpc url.

We are aware of that, and we are planning to change it in the future. For now, let's leave it as it is.

barabanovro commented:
Thank you for the review. I'll address the requested changes today

rjnrohit commented:
> Thank you for the review. I'll address the requested changes today

@barabanovro would you provide me collaborator access to the repository: https://github.com/walnuthq/starknet-foundry/tree/contract_verify, so that I can add a verification interface for voyager

barabanovro commented:
> > Thank you for the review. I'll address the requested changes today
> 
> @barabanovro would you provide me the collaborator access to the repository: https://github.com/walnuthq/starknet-foundry/tree/contract_verify, so that I can add a verification interface for voyager

This PR is almost complete and ready to be merged, which is essential for an event we are attending next week. To avoid any delays, we would prefer to keep this PR as is and not introduce additional complexity. Once it is merged, @rjnrohit, you can proceed with your verification. We hope this works for you!

ksew1 commented:
> A `package` variable must be added inside the `Verify` struct, refer `Declare` object

@barabanovro did you address this? I can't find that in code

rjnrohit commented on crates/sncast/src/main.rs:
```diff
@@ -432,7 +432,7 @@ async fn run_async_command(
         Commands::Verify(verify) => {
             let manifest_path = assert_manifest_path_exists()?;
-            let package_metadata = get_package_metadata(&manifest_path, &None)?;
+            let package_metadata = get_package_metadata(&manifest_path, &verify.package)?;
```
It'd be better if `&package_metadata.manifest_path` is passed instead of `&manifest_path` in the `verify function call `, [ref](https://github.com/foundry-rs/starknet-foundry/blob/b634b7a1a93b2cdd4d3865330a889c1ad5b4a011/crates/sncast/src/main.rs#L451)

barabanovro commented:
@ksew1 I have now addressed the comment and added the corresponding tests. Additionally, I have linked the appendix entry in `SUMMARY.md`.

Regarding the overview of this command under the `docs/src/starknet` directory, since creating a good doc will take some time, I'm wondering if I could add a proper overview in a separate PR. If it is necessary to include it in this PR, I can certainly add it here

Arcticae commented:
@barabanovro it's fine, you can add a guide in a next PR

ksew1 reviewed (approved):
LGTM

barabanovro commented:
@Arcticae What are the next steps to merge the PR?
